### PR TITLE
[Merged by Bors] - Add API to compute discrete validator attestation performance

### PR DESCRIPTION
**Pull request by macladson**
## Issue Addressed

N/A

## Proposed Changes

Add a HTTP API which can be used to compute the attestation performances of a validator (or all validators) over a discrete range of epochs.
Performances can be computed for a single validator, or for the global validator set. 

### Request
The API can be used as follows:
```
curl "http://localhost:5052/lighthouse/analysis/attestation_performance/{validator_index}?start_epoch=57730&end_epoch=57732"
```
Alternatively, to compute performances for the global validator set:
```
curl "http://localhost:5052/lighthouse/analysis/attestation_performance/global?start_epoch=57730&end_epoch=57732"
```

### Response
The response is JSON formatted as follows:
```
[
  {
    "index": 72,
    "epochs": {
      "57730": {
        "active": true,
        "head": false,
        "target": false,
        "source": false
      },
      "57731": {
        "active": true,
        "head": true,
        "target": true,
        "source": true,
        "delay": 1
      },
      "57732": {
        "active": true,
        "head": true,
        "target": true,
        "source": true,
        "delay": 1
      },
    }
  }
]
```
> Note that the `"epochs"` are not guaranteed to be in ascending order. 

## Additional Info

- This API is intended to be used in our upcoming validator analysis tooling (#2873) and will likely not be very useful for regular users. Some advanced users or block explorers may find this API useful however.
- The request range is limited to 100 epochs (since the range is inclusive and it also computes the `end_epoch` it's actually 101 epochs) to prevent Lighthouse using exceptionally large amounts of memory.

### CODE REVIEW

File: beacon_node/http_api/src/attestation_performance.rs
Comment (michaelsproul) [inline beacon_node/http_api/src/attestation_performance.rs]:
```suggestion
```

Can delete this

File: beacon_node/http_api/src/attestation_performance.rs
Comment (michaelsproul) [inline beacon_node/http_api/src/attestation_performance.rs]: Because the validator indices are only required in the `"global"` branch we can avoid locking the head or cloning here.

I think we can just get the number of validators in a `with_head` closure like this:

```rust
let num_validators = chain
  .with_head(|head| Ok(head.beacon_state.validators().len() as u64))
  .map_err(beacon_chain_error)?;
```

This can also get pushed down under the `if .. { }`

File: beacon_node/http_api/src/attestation_performance.rs
Comment (michaelsproul) [inline beacon_node/http_api/src/attestation_performance.rs]:
```suggestion
    // Load first block so we can get its parent.
```

File: beacon_node/http_api/src/attestation_performance.rs
Comment (michaelsproul) [inline beacon_node/http_api/src/attestation_performance.rs]: If we do the `|(root, _)| root` mapping before collecting the first time we can avoid some allocations here. It'll end up being a double map like:

```rust
.map(|res| res.map(|(root, _)| root))
```

**Review by michaelsproul — CHANGES_REQUESTED** [review]
Looks good on a first pass. Just a few perf improvements suggested :ok_hand:

File: beacon_node/http_api/src/attestation_performance.rs
Comment (macladson) [reply beacon_node/http_api/src/attestation_performance.rs]: Yeah good call!

File: beacon_node/http_api/src/attestation_performance.rs
Comment (macladson) [reply beacon_node/http_api/src/attestation_performance.rs]: I always do this :facepalm:

File: common/eth2/src/lighthouse/attestation_performance.rs
Comment (michaelsproul) [inline common/eth2/src/lighthouse/attestation_performance.rs:12]: let's add a `serde(skip_serializing_if = "Option::is_none")` to save outputting `null` values.

**Review by michaelsproul — APPROVED** [review]
Happy to merge modulo one smol optimisation

**Comment by michaelsproul** [comment]
Will merge once v2.1.0 is released.

File: beacon_node/http_api/src/attestation_performance.rs
Comment (michaelsproul) [inline beacon_node/http_api/src/attestation_performance.rs]: Sorry, me again :sweat_smile:

I noticed that we return a 500 internal server error if the `query.start_epoch` is `>= current_epoch - 1` due to the lookahead stuff. Would you mind adding a condition here like: `if query.end_epoch >= current_epoch - 1 { /* bad request */ }`. There are probably a few ways to write equivalent conditions but I think looking at the `end_epoch` makes sense. I think pretty much we need to just enforce that queries only include data up to the `current_epoch - 2`.

**Review by michaelsproul — CHANGES_REQUESTED** [review]

File: beacon_node/http_api/src/attestation_performance.rs
Comment (macladson) [reply beacon_node/http_api/src/attestation_performance.rs]: Oh nice catch!
I think I've fixed it now. 
I also adjusted some of the other error messages

**Review by michaelsproul — APPROVED** [review]
Let's do it!

bors r+

**Comment by michaelsproul** [comment]
I think we're blocked on CLA assistant working again (or getting disabled)

**Comment by michaelsproul** [comment]
[Re-check CLA](https://cla-assistant.io/check/sigp/lighthouse?pullRequest=2874)

**Pull request closed without merging**
